### ISSUE TITLE
Decouple layout and content

### ISSUE THREAD

**Issue by tylersticka** [cloudfour/drizzle-builder#14]
It's really bizarre to me that content pages like "overview," "components," etc. are nested within `views/`. These pages often include documentation content, yet cannot use Markdown and exist separately from "docs." It feels strange to me that the only way to create a new, non-Markdown page at the root or an arbitrary location is to create a brand-new "view."

I guess instead of this:

```
├── docs/
│   └── javascript.md    → /docs/javascript.html
└── views/
    ├── layouts/
    ├── index.html       → /index.html
    └── typography.html  → /typography.html
```

I'd love for the structure to be more like this (actual directory names aren't as important as the structure more clearly influencing the end result):

```
├── content/
│   ├── docs/
│   │   └── javascript.md  → /docs/javascript.html
│   ├── index.html         → /index.html
│   └── typography.md      → /typography.html
└── layouts/
```

**Comment by lyzadanger** [comment]
All righty, it is now time to answer the question, based on our ideal situation:

_What defines what HTML pages get created?_

I've heard input that a hard coupling between `views` and directories full of `patterns` is not ideal. Now's your time to brainstorm! What do you think is the best logic for determining the structure of the output HTML content?

This includes all of the pieces involved: docs, patterns (and collections of patterns), sandbox and any arbitrary pages you may wish to add to the mix.

**Comment by tylersticka** [comment]
Okay, this is truly a _brainstorm_ for me. But I'm going to describe how _I_ might structure this. I am not married to what I describe.

## Types of Stuff

### Layouts (or "Templates"?)

Layouts are written in Handlebars (or whatever). They do not represent the compiled results. They are simply the containers for optional page content.

As an example, consider a layout file named `simple.hbs` that looks like this:

```hbs
<!doctype html>
<html>
<head>
  <title>{{title}}</title>
</head>
<body>
  {{{contents}}}
</body>
</html>
```

This would not result in a `simple.html` page being compiled. But any page (more on that later) that specified `layout: simple` in its Frontmatter would be compiled with the aforementioned layout.

### Pages (or "Content"?)

Pages define static content to be compiled. They should have a default `layout` assigned to them so you don't need to specify, but you can override the `layout` via Frontmatter.

This would allow you to structure any informational pages, documentation, etc. any way you'd like:

```
├── docs/
│   └── javascript.md    → /docs/javascript.html
├── index.md             → /index.html
├── pages/
│   ├── home.hbs         → /pages/home.html
│   └── product.hbs      → /pages/product.html
└── sandbox/
    └── my-prototype.hbs → /sandbox/my-prototype.html
```

If you wanted a page to _not_ have the style guide nav, or to use some completely different layout, you'd set the `layout` property within its Frontmatter.

### Patterns and Collections

The most complicated part of things are the _patterns_ themselves. Which kinda makes sense... that's the whole reason we're making a tool, and not simply re-using a static site generator!

Directories in `patterns/` result in compiled pages as long as they have one or more _visible_ pattern files as immediate children:

```
├── typography/               → /typography.html
│   ├── headings.hbs
│   └── paragraph.hbs
├── components/
│   ├── button/               → /button.html
│   │   ├── 00-base.hbs
│   │   ├── 01-primary.hbs
│   │   ├── 02-secondary.hbs
│   │   └── social.hbs
│   └── icon/                 → (nothing, includes only a hidden pattern)
│       └── _icon.hbs
└── utilities/                → /utilities.html
    ├── space.hbs
    └── text.hbs
```

Patterns (including hidden) are available to other patterns and pages as partials (`{{>collection.pattern}}`).

Layouts have access to two objects that result from this structure:
- An object representing the hierarchy of all visible patterns. This allows layout templates to construct navigation.
- An object representing the _current_ collection (if any). This can be used by the layout that compiles the collection pages.

Individual patterns are never compiled. They must always be part of a collection, or included by a page.

## Putting It All Together

I find it helps sometimes to visualize what the _start_ and _end_ results might be. Here's my attempt to visualize an example source and destination directories (for this stuff only, not assets and the like):

### Source

```
├── layouts/
│   ├── partials/
│   │   ├── header.hbs
│   │   ├── footer.hbs
│   │   └── nav.hbs
│   ├── default.hbs
│   ├── page.hbs
│   └── sandbox.hbs
├── pages/
│   ├── docs/
│   │   └── javascript.md
│   ├── index.md
│   ├── pages/
│   │   └── home.hbs
│   └── sandbox/
│       └── my-prototype.hbs
└── patterns/
    ├── typography/
    │   ├── headings.hbs
    │   └── paragraph.hbs
    ├── components/
    │   ├── button/
    │   │   ├── 00-base.hbs
    │   │   ├── 01-primary.hbs
    │   │   ├── 02-secondary.hbs
    │   │   └── social.hbs
    │   └── icon/
    │       └── _icon.hbs
    └── utilities/
        ├── space.hbs
        └── text.hbs
```

### Result

```
├── button.html
├── docs/
│   └── javascript.html
├── index.html
├── pages/
│   └── home.html
├── sandbox/
│   └── my-prototype.html
├── typography.html
└── utilities.html
```

## Misc. Thoughts

- I'd love if files with the extension `.md` were parsed with Markdown, while files with a Handlebars extension were compiled with handlebars. This would be the most helpful with pages (a documentation page could end in `.md` for instance), but it might also be useful for typographic patterns?
- I don't really care if `patterns/components/button/` gets compiled to `button.html` or `components/button.html`.
- If we do have example pages, it might be confusing for them to be nested in `pages/pages` in our source directory..?

**Comment by lyzadanger** [comment]
Riffing on what you wrote:

* I think it's nice to have a directory that, _as closely as possible_, contains things that will result in output HTML. In my case, I'm grouping most of this under a `content` directory. Any single file under `docs`, `pages`, or `sandbox` will result in an output HTML file. It's not as easy to do a one-to-one representation in source of the `patterns`, but I think that's OK.
* I'm segregating `templates` outside of `content` and sticking `layouts` and `partials` inside.
* One would be able to dictate within `frontMatter` what `layout` or whatnot to use against a given resource within `content` (excepting `patterns`), with some sane defaults (`page.hbs`, `docs.hbs`, e.g.).
* I'm still having some semantic difficulty between `layouts` and, uh, not-layout `hbs` source. 

## Source

```
├── content
│   ├── docs
│   │   ├── 01-howto.md
│   │   └── javascript.md
│   ├── pages
│   │   ├── arbitrary-page.hbs
│   │   └── nested-page
│   │       └── index.hbs
│   ├── patterns
│   │   ├── components
│   │   │   ├── button
│   │   │   │   ├── 00-base.hbs
│   │   │   │   ├── 01-primary.hbs
│   │   │   │   ├── 02-secondary.hbs
│   │   │   │   └── social.hbs
│   │   │   ├── icon
│   │   │   │   └── _icon.hbs
│   │   │   └── link
│   │   │       └── 01-visited.hbs
│   │   ├── typography
│   │   │   ├── headings.hbs
│   │   │   └── paragraph.hbs
│   │   └── utilities
│   │       ├── space.hbs
│   │       └── text.hbs
│   └── sandbox
│       └── my-prototype.hbs
└── templates
    ├── default.hbs
    ├── docs.hbs
    ├── layouts
    ├── page.hbs
    ├── partials
    │   ├── footer.hbs
    │   ├── header.hbs
    │   └── nav.hbs
    └── sandbox.hbs
```

Resulting in:

## Output

```
├── arbitrary-page.html
├── docs
│   ├── 01-howto.html
│   └── javascript.html
├── nested-page
│   └── index.html
├── patterns
│   ├── components
│   │   ├── button.html
│   │   └── link.html
│   ├── index.html
│   ├── typography.html
│   └── utilities.html
└── sandbox
    └── my-prototype.html
```

Note: 

* Nesting of the pattern library under `patterns`. I suggest this because it's easier to scan, lets us potentially use the output for more than a pattern library, and prevents namespace collisions.
* For pattern output, I'm going here with an approach of: any directory under `patterns` in source that has any pattern files in it will generate an HTML page. That HTML page would be named after the parent directory and would contain the patterns from that directory. I do like nesting the output directories so that they represent source (not flattening to top level). Note that a patterns subdirectory that only contained directories would not result in an HTML file.
* Anything under `pages` in source would get output at the top level.

It's the end of the day so I might be getting a little fuzzy...but this feels like it's overall going the right direction! Keep it coming!

**Comment by tylersticka** [comment]
In your example tree, I'm not sure why `docs/` are not handled just like `pages/`? I know in Fabricator's example they are handled as "material," but in our projects we never use them that way. I think they should just be pages, and we should allow pages to be written in Markdown (`page-title.md`).

> Nesting of the pattern library under patterns. I suggest this because it's easier to scan, lets us potentially use the output for more than a pattern library, and prevents namespace collisions.

I am not emphatically opposed to this. On principle, it's really no different than nesting `journal/` or `posts/` under `content/` and treating it special in a static site (a strategy I believe you and I both employ).

That said, I can't help but wonder if separating `patterns/` from `pages/` might be useful as a cognitive line in the sand, a clear indication that `patterns/` are handled differently than `pages/`. In Fabricator, it's confusing that `layouts/` live together with `views/`... I'm not sure it would be any less confusing to have `material/` living in `views/` instead.

I did consider this sort of grouping while writing my previous comment:

```
├── content/
│   ├── pages/
│   └── patterns/
└── templates/
```

I went back on it because it felt strange to nest the thing I spend about 30-50% of my time editing (`patterns/`) deeper than something I rarely edit `templates/`. But by that logic, our Gulpfile should be buried somewhere bizarre and all our patterns should be in the repo root, so I'm unsure how valid an argument that is! :laughing:

**Comment by lyzadanger** [comment]
Briefly, I agree with most of your feedback. I think your insight about having `patterns` elsewhere because they _don't_ have a one-to-one pages-output connection is a good one. Let me think on the docs one—it may be fine; something is bugging me about it but I haven't quite figured out what yet :). More soon.

**Comment by tylersticka** [comment]
As more food for thought re: docs...

I've always hated the heading "docs," because the whole deliverable is documentation. It doesn't really _say_ anything to the viewer.

If you look at our most recent shipped pattern library (no link since it's not open source), it seems arbitrary to me that a page like "Icons" is treated differently than a page like "Naming Conventions." They're both pages documenting how to do things with a combination of prose and figures/examples... the only difference is _where_ they're output, and whether or not they're written in Markdown or HTML.

Also using that pattern library as an example, it seems completely silly that a page like "Overview" absolutely _cannot_ be written in Markdown. It's 100% prose! :laughing:

**Comment by erikjung** [comment]
> TL;DR: I feel like relegating documentation in Markdown to a specific, special directory inadvertently establishes a "technical documentation ghetto." I'd like it to be a first-class citizen.

I agree with this. It seems like Markdown should be able to exist anywhere, and by default should be rendered to a corresponding HTML page with an assumed layout that can be overridden via front-matter.

```
patterns
└── components
    └── button
        ├── 00-base.hbs
        ├── 01-primary.hbs
        ├── 02-secondary.hbs
        └── index.md
```

```
---
foo: bar
baz: bang
layout: some-other-layout
---

This is the content. Maybe this can take the place of "notes"?

Also worth noting:

<div>
  <p>We can always still have regular HTML within Markdown.</p>
</div>
```

**Comment by lyzadanger** [comment]
I would propose this:

* Any file under `pages` that is `.md` will be parsed as markdown and will generate an output HTML file
* Let's not do this inside of the `patterns` structure—it's a bit too hard to scan what input/output relationships are for files both for humans and for the already-complex globbing and parsing mechanism for `patterns`

That is, yep, let's support markdown-anywhere inside of the content directories, but let's not do it from inside `patterns` as it is already complicated enough in there.

Thoughts?

**Comment by lyzadanger** [comment]
Expanding a little:

> Any file...that is .md will be parsed as markdown and will generate an output HTML file

By _default_, that's how we would treat those files. Let's pretend we call this feature (processing content and generating HTML output) `docPages` (though we probably won't because that term is lame). Say the default handling is something like (this is pseudo-code):

```
const defaults = {
// ...
  docPages: {
    glob:  'src/content/pages/**/*.md',
    contentFn: (contents, filePath) => marked(contents),
    defaultLayout: 'src/templates/docs.hbs'
  }
};
```

The notion of `contentFn` is something I'm already using in a few places—it provides a sensible default function for running over files' contents when they're parsed (e.g. `JSON` over JSON data, `yaml` over YAML data) but can be overridden with options. In this case, we'd have a default content-parsing function of `markdown` and look for `.md` files, but there's no reason the glob or the parsing function couldn't be specified to do something differently in a given installation.

The implication here is that `defaultLayout` (or whatever we ultimately name it) could be overridden in front matter.

**Comment by tylersticka** [comment]
> Let's not do this inside of the patterns structure—it's a bit too hard to scan what input/output relationships are for files both for humans and for the already-complex globbing and parsing mechanism for patterns

If this is a _technical_ limitation, I'm fine with it. I'm not so sure about it as a _design_ decision. Some patterns, particularly those describing default typography, might be easier to write using Markdown.

**Comment by lyzadanger** [comment]
> If this is a technical limitation, I'm fine with it. I'm not so sure about it as a design decision. Some patterns, particularly those describing default typography, might be easier to write using Markdown.

It's technical if we want `.md` files inside of `patterns` to generate a _unique output page_. It's not a problem as much if we want `.md` files inside of `patterns` to get `marked`-parsed but otherwise be treated as a pattern.

**Comment by tylersticka** [comment]
> It's not a problem as much if we want .md files inside of patterns to get marked-parsed but otherwise be treated as a pattern.

:point_up: This is what we want for sure!

**Comment by lyzadanger** [comment]
All right! An attempt to boil down the things I've heard and seen and thought! Let's propose this as a `src` tree:

```
├── pages
│   ├── about.hbs
│   ├── about.md
│   ├── arbitrary-page.hbs
│   ├── docs
│   │   ├── 01-howto.md
│   │   └── javascript.md
│   ├── nested-page
│   │   └── index.hbs
│   ├── sandbox
│   │   └── my-prototype.hbs
│   └── somePage.md
├── patterns
│   ├── components
│   │   ├── button
│   │   │   ├── 00-base.hbs
│   │   │   ├── 01-primary.hbs
│   │   │   ├── 02-secondary.hbs
│   │   │   └── social.hbs
│   │   ├── icon
│   │   │   └── _icon.hbs
│   │   └── link
│   │       ├── 01-visited.hbs
│   │       └── notes-about-links.md
│   ├── thing.md
│   ├── typography
│   │   ├── headings.hbs
│   │   └── paragraph.hbs
│   └── utilities
│       ├── space.hbs
│       └── text.hbs
└── templates
    ├── layouts
    │   └── default.hbs
    ├── page.hbs
    ├── partials
    │   ├── footer.hbs
    │   ├── header.hbs
    │   └── nav.hbs
    └── sandbox.hbs
```

And as output:

```
├── about-01.html
├── about.html
├── arbitrary-page.html
├── docs
│   ├── 01-howto.html
│   └── javascript.html
├── nested-page
│   └── index.html
├── patterns
│   ├── components
│   │   ├── button.html
│   │   └── link.html
│   ├── patterns.html
│   ├── typography.html
│   └── utilities.html
└── sandbox
    └── my-prototype.html
```

Now to talk through how we got to this `dist`:

## Pages

* There is a one-to-one relationship for all files under `pages` to a single `HTML` file.
* Per suggestion, we can parse the contents of these files differently depending on extension.
* Pages, like other files, have recourse to `frontMatter` to override defaults and set data.
* Note that I've introduced a purposeful namespace collision by creating an `about.hbs` and an `about.md` file. Let me know if there's a different way you'd like to handle that than what I've done.
* I've left a `docs` directory there but that name doesn't have any real meaning. Point being: you can nest pages however you like.

## Patterns

* In this example, I've put patterns inside of a `patterns` directory in `dist`. I kind of like that, but am also amenable to making that an option that can be passed to the builder.
* The rule for creating an HTML page from anything in `patterns`: If a directory has any files in it, a page, named after the directory, will be created. Ergo `patterns.html` but not `components.html`. (`Patterns.html` would contain the `thing.md` output).

I still need to get a sense of how `sandbox` items differ from other `pages`. I'll get to that at some point!

/cc @tylersticka @erikjung @mrgerardorodriguez

**Comment by tylersticka** [comment]
> Note that I've introduced a purposeful namespace collision by creating an about.hbs and an about.md file. Let me know if there's a different way you'd like to handle that than what I've done.

I almost think this should just error out or show a warning and override the file. I would rather the tool tell me "hey, these aren't named uniquely" than to inadvertently start relying on permalinks that were auto-renamed like that. That's just my knee-jerk opinion!

> In this example, I've put patterns inside of a patterns directory in dist. I kind of like that, but am also amenable to making that an option that can be passed to the builder.

I'd prefer if they were just output to `dist` directly. I feel like the "patterns" distinction is only relevant during development, it's not really super relevant to the end-user (someone viewing the library). But I agree this should be an option.

> I still need to get a sense of how sandbox items differ from other pages. I'll get to that at some point!

I'm not sure they need to differ? As long as pages can define a new `layout` (or `template`) to use (one that's essentially "blank" with none of our style guide UI), that's all we'll need for `sandbox`.

**Comment by lyzadanger** [comment]
I'm a call this done.
